Add descriptions of ref conditional expression

BillWagner opened the pull request:
Fixes #6534

/cc @PauloMorgado

- Update initial description of ?:.
- Add remarks that the conditions for `ref` results must be the same. An implicit conversion is not sufficient.
- Add example of ref expression
- Move samples to sample repository. (See dotnet/samples#432)

Relies on dotnet/samples#432

BillWagner committed:
Add descriptions of ref conditional expression

Fixes #6534

/cc @PauloMorgado

- Update initial description of ?:.
- Add remarks that the conditions for `ref` results must be the same. An implicit conversion is not sufficient.
- Add example of ref expression
- Move samples to sample repository. (See dotnet/samples#432)

pkulikov commented on docs/csharp/whats-new/csharp-7-2.md:
```diff
@@ -60,9 +62,19 @@ int binaryValue = 0b_0101_0101;
+
+## Conditional `ref` expressions
+
+Finally, the conditional expression may produce a ref result instead of a value result. For example, you would write the following to retrieve a reference to one of two arrays:
```
"reference to one of two arrays" -> "reference to the first element of one of two arrays"

pkulikov commented on docs/csharp/whats-new/csharp-7-2.md:
````diff
@@ -60,9 +62,19 @@ int binaryValue = 0b_0101_0101;
+Finally, the conditional expression may produce a ref result instead of a value result. For example, you would write the following to retrieve a reference to one of two arrays:
+
+```csharp
+ref var r = ref (arr != null ? ref arr[0]: ref otherArr[0]);
````
nit: a white space before colon is missing.

BillWagner committed:
respond to feedback.

Thraka commented on docs/csharp/language-reference/operators/conditional-operator.md:
```diff
@@ -10,20 +10,31 @@ helpviewer_keywords:
+
+The `condition` must evaluate to `true` or `false`. If `condition` is `true`, `first_expression` is evaluated and becomes the result. If `condition` is `false`, `second_expression` is evaluated and becomes the result. Only one of the two expressions is evaluated.
+
+The type of `first_expression` and `second_expression` must be the same. If the conditional expression result is a value (not a `ref`), the expression compiles and executes if there exists an implicit conversion one type to the other.
```
I don't feel like this specifically calls out that ref doesn't support implicit conversion. (at least that is what your issue summary says)

BillWagner committed:
respond to feedback

Be more clear on identity and implicit conversions for ref results.

Add note on the importance that only one of the result expressions is evaluated.

BillWagner commented:
closing and reopening to start a new build

BillWagner committed:
fix build error

rpetrusha commented on docs/csharp/language-reference/operators/conditional-operator.md:
````diff
@@ -10,20 +10,37 @@ helpviewer_keywords:
-  
+```
+
+Beginning with C# 7.2, the `first_expression` and `second_expression` my be `ref` expressions:
````
It would be helpful to make ref expressions a link to the C# topic.

BillWagner replied:
I haven't been doing that because the spec we host is the C# 6 spec. Should I link to the design proposal in the dotnet/csharplang repo? 

What are your thoughts?

rpetrusha replied:
I think that's fine.

rpetrusha commented on docs/csharp/language-reference/operators/conditional-operator.md:
````diff
@@ -10,20 +10,37 @@ helpviewer_keywords:
+ref condition ? ref first_expression : ref second_expression;  
+```
+
+The result may be assigned to a `ref` or `ref readonly` variable, or a variable with neither modifier.
````
nit: or a --> or to a

rpetrusha commented on docs/csharp/language-reference/operators/conditional-operator.md:
```diff
@@ -10,20 +10,37 @@ helpviewer_keywords:
+
+## Remarks
+
+The `condition` must evaluate to `true` or `false`. If `condition` is `true`, `first_expression` is evaluated and becomes the result. If `condition` is `false`, `second_expression` is evaluated and becomes the result. Only one of the two expressions is evaluated. This is particularly important for expressions where the result is a `ref` as the following is valid:
```
comma before "as the following"

rpetrusha commented on docs/csharp/whats-new/csharp-7-2.md:
````diff
@@ -60,9 +62,19 @@ int binaryValue = 0b_0101_0101;
+ref var r = ref (arr != null ? ref arr[0] : ref otherArr[0]);
+```
+
+The variable `r` is a reference to the first value in either `arr` or `otherArr`. For more information, see [conditional operator (?:)](../language-reference/operators/conditional-operator.md) in the language reference.
````
"For more information..." should begin a second paragraph.

rpetrusha reviewed (approved):
I've left a number of nits, @BillWagner. You decide whether you want to address them and then merge.

BillWagner committed:
C# scripting doesn't support ref yet.

So don't use it.

BillWagner committed:
respond to feedback.

BillWagner committed:
add link to C# feature speclet.
